src/docfx/build/toc/TableOfContentsMap.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace Microsoft.Docs.Build
+{
+    /// <summary>
+    /// The mappings between toc and document
+    /// </summary>
+    internal class TableOfContentsMap
+    {
+        private readonly IEnumerable<Document> _tocs;
+
+        private readonly Dictionary<Document, HashSet<Document>> _documentToTocs;
+
+        private readonly ConcurrentDictionary<Document, Document> _nearestTocs = new ConcurrentDictionary<Document, Document>();
+
+        public TableOfContentsMap(IEnumerable<Document> tocs, Dictionary<Document, HashSet<Document>> documentToTocs)
+        {
+            _tocs = tocs ?? throw new ArgumentNullException(nameof(tocs));
+            _documentToTocs = documentToTocs ?? throw new ArgumentNullException(nameof(documentToTocs));
+        }
+
+        /// <summary>
+        /// Find the toc relative path to document
+        /// </summary>
+        /// <param name="file">Document</param>
+        /// <returns>The toc relative path</returns>
+        public string FindTocRelativePath(Document file)
+        {
+            var nearestToc = _nearestTocs.GetOrAdd(file, add => GetNearestToc(add));
```

Comment:
this extra caching is unnessessary

Reply:
you are right

---
In reply to: [184583269](https://github.com/dotnet/docfx/pull/2696#discussion_r184583269) [](ancestors = 184583269)